### PR TITLE
Immutable record default values utilities

### LINKED ISSUE #129: AuditableDefaultValues / Factory for requiring properties to be set
Add a constant object that fills out default values for the `Auditable` interface, which are commonly setup for immutable records (otherwise, those properties won't be populated):

Example of current process:
```ts
const defaultValues: OfflineDevice = {
    createdById: undefined,
    createdOn: undefined,
    deletedById: undefined,
    deletedOn: undefined,
    deviceId: "",
    id: undefined,
    name: "",
    offlineBooks: [],
    updatedById: undefined,
    updatedOn: undefined,
};

export class OfflineDeviceRecord extends Record(defaultValues) implements OfflineDevice
// ...
```

Proposed example API:
```ts
import { AuditableDefaultValues } from "andculturecode-javascript-core";

const defaultValues: OfflineDevice = {
    ...AuditableDefaultValues
    deviceId: "",
    name: "",
    offlineBooks: [],
};
```

A factory function which would require all of the properties of `T` to be defined would also be nice. It is a common pitfall for engineers to forget a property in the `defaultValues` object, which will then not be populated even when manually set during an API request for example.

The above could look something like this (quick scratch code - not 100% sure if this will work)

```ts
// Somewhere in andculturecode-javascript-core...
const makeDefaultValues = <T>(values: Required<T>): T => values as T;
const withAuditableDefaultValues = <T extends Auditable>(
  values: Required<Exclude<T, Auditable>>
): T => Object.assign(values, DefaultAuditableValues) as T;

import {
  Auditable,
  DefaultAuditableValues,
  makeDefaultValues,
  withAuditableDefaultValues,
} from "andculturecode-javascript-core";

const defaultValues: OfflineDevice = makeDefaultValues<OfflineDevice>(
  withAuditableDefaultValues({
    deviceId: "",
    name: "",
    offlineBooks: [],
  })
);
```

### PR DESCRIPTION
**Pull request by brandongregoryscott**
Resolves #129 AuditableDefaultValues / Factory for requiring properties to be set

- `defaultValuesFactory` which essentially just provides typing to ensure consumer doesn't miss any properties of type `T`
- `auditableDefaultValuesFactory` which does the same, but also provides defined properties for `Auditable`s
- `ensureRecords` which takes an array of objects and maps it through `ensureRecord`, which is common for view model constructors to do for navigation properties
- Some additional cleanup, dependency bumps and removal of ported types

---

- [x] Related GitHub issue(s) linked in PR description
- [x] Destination branch merged, built and tested with your changes
- [x] Code formatted and follows best practices and patterns
- [x] Code builds cleanly (no _additional_ warnings or errors)
- [x] Manually tested
- [x] Automated tests are passing
- [x] No _decreases_ in automated test coverage
- [x] Documentation updated (readme, docs, comments, etc.)
- [-] Localization: No hard-coded error messages in code files (_minimally_ in string constants)

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1,8 +1,11147 @@
 {
     "name": "andculturecode-javascript-core",
     "version": "0.6.4",
-    "lockfileVersion": 1,
+    "lockfileVersion": 2,
```
Comment (brandongregoryscott) [inline package-lock.json:4]: Looks like the massive change here was from updating `npm` to version 7 on my machine. V2 lockfiles should be backwards compatible with npm v5/6 so this shouldn't _really_ be an issue, but I can certainly revert it and regenerate it with an older version

File: src/utilities/record-utils.ts [hunk 2]
```diff
@@ -21,6 +46,21 @@ const RecordUtils = {
...
+    /**
+     * Function to ensure a given array of objects are instances of a specific Record `T`. If they
+     * are not, one will be instantiated with the given constructor.
+     *
+     * If `maybeRecord` is already an instance of `T`, it returns that value.
+     *
+     * @param maybeRecords Array of objects or Records to be coalesced into Records.
+     * @param record Type of the Record to be instantiated
+     */
+    ensureRecords<T>(maybeRecords: any[], record: Constructor<T>): T[] {
```
Comment (myty) [inline src/utilities/record-utils.ts:58]: Thought: What if this was more strongly typed?  Something like this: 
```
ensureRecords<TValue, TRecord extends TValue>(
        maybeRecords: TValue[],
        record: Constructor<TRecord>
    ): TRecord[]
```

**Review by myty — APPROVED** [review]
Looks good!

File: src/utilities/record-utils.ts (hunk 2 above)
Comment (brandongregoryscott) [reply src/utilities/record-utils.ts:58]: I don't see why it couldn't be - let me try it out 👍

**Review by dylanjustice — APPROVED** [review]

**Comment by dylanjustice** [comment]
Love the `ensureRecords` addition. That will come in handy.

**Comment by brandongregoryscott** [comment]
@all-contributors add @dylanjustice for review

**Pull request merged**